rir/src/compiler/opt/loop_invariant.cpp

```diff
@@ -0,0 +1,191 @@
+#include "../pir/pir_impl.h"
+#include "../transform/bb.h"
+#include "../util/cfg.h"
+#include "../util/safe_builtins_list.h"
+#include "../util/visitor.h"
+#include "R/r.h"
+#include "pass_definitions.h"
+
+namespace rir {
+namespace pir {
+
+typedef std::pair<BB*, BB*> BackEdge;
+typedef std::function<bool(Instruction*)> InstrActionPredicate;
+
+auto noEnvironmentTainting = [](Instruction* i) {
+    if (LdFun::Cast(i))
+        return true;
+
+    if (auto call = CallBuiltin::Cast(i))
+        return SafeBuiltinsList::nonObject(call->builtinId);
+
+    if (CallSafeBuiltin::Cast(i))
+        return true;
+
+    if (CallInstruction::CastCall(i))
+        return false;
+
+    return !(i->mayUseReflection());
+};
+
+struct NaturalLoop {
+    BB* header;
+    std::vector<BB*> tails; // nodes that precede the header
+    std::unordered_set<BB*> body;
+
+    bool isSafeToHoistLoads() { return run(noEnvironmentTainting); }
+
+    bool overwrites(SEXP binding) {
+        return !run([&](Instruction* i) {
+            // A store overwrites the binding
+            SEXP varName = nullptr;
+            if (auto store = StVar::Cast(i))
+                varName = store->varName;
+            else if (auto store = StVarSuper::Cast(i))
+                varName = store->varName;
+
+            if (varName && varName == binding)
+                return false;
+
+            // An environment overwrites the binding
+            if (auto env = MkEnv::Cast(i)) {
+                auto safe = true;
+                env->eachLocalVar([&](SEXP name, Value* v) {
+                    if (name == binding) {
+                        safe = false;
+                        return;
+                    }
+                });
+                if (!safe)
+                    return false;
+            }
+
+            return true;
+        });
+    }
+
+    bool run(const InstrActionPredicate& action) {
+        auto answer = true;
+        for (auto bb : body) {
+            for (auto instruction : *bb) {
+                if (!action(instruction)) {
+                    answer = false;
+                    break;
+                }
+            }
+        }
+        return answer;
+    }
+
+    BB* outOfLoopPredecessor(const CFG& cfg) {
+        std::vector<BB*> diff;
+        std::set_difference(cfg.immediatePredecessors(header).begin(),
+                            cfg.immediatePredecessors(header).end(),
+                            body.begin(), body.end(),
+                            std::inserter(diff, diff.end()));
+
+        return diff.front();
```

Comment:
I don't think this is right? What if there are multiple (non-loop) nodes that precede the header?

I think we want the _loop preheader_, which is the single node that's the only way to enter the loop. It doesn't always exist, so sometimes you have to create it. But the point is that anything hoisted out of the loop should go there.

I'll see if I can add it to the loop API.

Reply:
Yes, definetely. There was an assert in the first version I do not know why I removed it. Ping me if you add the code for the preheader.